save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/DownloadFilesController.kt

```diff
@@ -32,25 +36,38 @@ import kotlin.io.path.*
  * A Spring controller for file downloading
  */
 @RestController
+@Suppress("LongParameterList")
 class DownloadFilesController(
     private val additionalToolsFileSystemRepository: TimestampBasedFileSystemRepository,
     private val testDataFilesystemRepository: TestDataFilesystemRepository,
     private val agentRepository: AgentRepository,
     private val organizationService: OrganizationService,
     private val userDetailsService: UserDetailsService,
+    private val projectService: ProjectService,
+    private val projectPermissionEvaluator: ProjectPermissionEvaluator,
 ) {
     private val logger = LoggerFactory.getLogger(DownloadFilesController::class.java)
 
     /**
      * @param organizationName
      * @param projectName
+     * @param authentication
      * @return a list of files in [additionalToolsFileSystemRepository]
      */
     @GetMapping(path = ["/api/$v1/files/{organizationName}/{projectName}/list"])
+    @Suppress("UnsafeCallOnNullableType")
     fun list(
         @PathVariable organizationName: String,
         @PathVariable projectName: String,
-    ): List<FileInfo> = additionalToolsFileSystemRepository.getFileInfoList(ProjectCoordinates(organizationName, projectName))
+        authentication: Authentication,
+    ): Mono<List<FileInfo>> = projectService.findWithPermissionByNameAndOrganization(
+        authentication, projectName, organizationName, Permission.WRITE
+    ).filter {
+        val project = projectService.findByNameAndOrganizationName(projectName, organizationName)
+        projectPermissionEvaluator.hasPermission(authentication, project!!, Permission.WRITE)
```

Comment:
I think this check is already performed isnide `findWithPermissionByNameAndOrganization`?

Reply:
Yes, agree. Fixed